staticaddr/script/script_test.go

```diff
@@ -0,0 +1,338 @@
+
+	testCases := []struct {
+		name    string
+		witness func(*testing.T) wire.TxWitness
+		valid   bool
+	}{
+		{
+			"success case coop spend with combined internal key",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = 1
+
+				// This is what gets signed.
+				taprootSigHash, err := txscript.CalcTaprootSignatureHash(
+					sigHashes, txscript.SigHashDefault, tx, 0,
+					prevOutFetcher,
+				)
+				require.NoError(t, err)
+
+				var msg [32]byte
+				copy(msg[:], taprootSigHash)
+
+				tweak := &input.MuSig2Tweaks{
+					TaprootTweak: staticAddress.RootHash[:],
+				}
+
+				sig, err := MuSig2Sign(
+					version, privKeys, pubKeys, tweak, msg,
+				)
+				require.NoError(t, err)
+
+				witness, err := staticAddress.GenSuccessWitness(
+					sig,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			"success case timeout spend with client key",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = uint32(csvExpiry)
+
+				sig, err := txscript.RawTxInTapscriptSignature(
+					tx, sigHashes, 0, value,
+					staticAddressScript,
+					*staticAddress.TimeoutLeaf,
+					txscript.SigHashAll, clientPrivKey,
+				)
+				require.NoError(t, err)
+
+				witness, err := staticAddress.GenTimeoutWitness(
+					sig,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, true,
+		},
+		{
+			"error case timeout spend with client key, wrong " +
+				"sequence",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = uint32(csvExpiry - 1)
+
+				sig, err := txscript.RawTxInTapscriptSignature(
+					tx, sigHashes, 0, value,
+					staticAddressScript,
+					*staticAddress.TimeoutLeaf,
+					txscript.SigHashAll, clientPrivKey,
+				)
+				require.NoError(t, err)
+
+				witness, err := staticAddress.GenTimeoutWitness(
+					sig,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+		{
+			"error case timeout spend with server key, server " +
+				"cannot claim timeout path",
+			func(t *testing.T) wire.TxWitness {
+				tx.TxIn[0].Sequence = uint32(csvExpiry)
+
+				sig, err := txscript.RawTxInTapscriptSignature(
+					tx, sigHashes, 0, value,
+					staticAddressScript,
+					*staticAddress.TimeoutLeaf,
+					txscript.SigHashAll, serverPrivKey,
+				)
+				require.NoError(t, err)
+
+				witness, err := staticAddress.GenTimeoutWitness(
+					sig,
+				)
+				require.NoError(t, err)
+
+				return witness
+			}, false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+
+		t.Run(testCase.name, func(t *testing.T) {
+			tx.TxIn[0].Witness = testCase.witness(t)
+
+			newEngine := func() (*txscript.Engine, error) {
+				return txscript.NewEngine(
+					staticAddressScript, tx, 0,
+					txscript.StandardVerifyFlags, nil,
+					sigHashes, value, prevOutFetcher,
+				)
+			}
+
+			assertEngineExecution(t, testCase.valid, newEngine)
+		})
+	}
+}
+
+func MuSig2Sign(version input.MuSig2Version, privKeys []*btcec.PrivateKey,
```

Comment:
I wonder if it's worth moving this to say `utils` package?

Reply:
There's a `utils.go` and a `utils/htlc_utils.go`. Should we create `utils/musig.go`. Although sometimes utils is an anti pattern.

Reply:
Yeah sgtm, it is somewhat of a too generic name I agree, but at least it is reusable.

Reply:
Ideally the server would also use this functionality.